mne/preprocessing/_csd.py

```diff
@@ -152,14 +189,17 @@ def compute_current_source_density(inst, sphere='auto', lambda2=1e-5,
     if not np.isfinite(pos).all() or np.isclose(pos, 0.).all(1).any():
         raise ValueError('Zero or infinite position found in chs')
     pos -= (x, y, z)
-
-    G = _calc_g(np.dot(pos, pos.T), stiffness=stiffness,
-                n_legendre_terms=n_legendre_terms)
-    H = _calc_h(np.dot(pos, pos.T), stiffness=stiffness,
-                n_legendre_terms=n_legendre_terms)
+    # Original code does not project onto the sphere, probably not good:
+    # pos /= radius
+    # from scipy.spatial.distance import squareform, pdist
+    # cos_dist = 1 - squareform(pdist(pos, 'sqeuclidean')) / 2.
```

Comment:
@alexrockhill I think this formulation might be incorrect (ignoring for now whether you need `pos /= radius` or `pos /= np.linalg.norm(pos, axis=1)`, let's assume you have `radius` right and it's actually spherical so they are the same). The [cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity) is given by the dot product divided by the magnidtudes (which are both one here). I haven't read the actual papers (probably should), but assuming that it's supposed to measure the cosine distance, using some variant of the squared Euclidean distance will not be correct.

Reply:
> [X,Y,Z] = sph2cart(ThetaRad,PhiRad,1.0)

So this takes the locations (wherever they were) and implicitly projects them onto the unit sphere. So we shouldn't do `pos /= radius` but rather `pos /= np.linalg.norm(pos, axis=1, keepdims=True)`. The remaining question in my mind is whether the 1 minus half the squared Euclidean distance is the correct thing to use or the cosine similarity between the positions. I'll read the paper to figure it out, but to me (given the spherical projection and variable naming) I'm guessing it will be the cosine similarity.

Reply:
one minus half the squared euclidean appears to be an equivalent way of computing the cosine similarity, so this is solved